File: ts/tests/adapters/fs.test.ts
import yaml from "js-yaml";
import { describe, expect, test } from "vitest";
import {
  FsAdapter,
  chainNormalizers,
  type FsRequest,
  type PathNormalizer,
} from "../../src/adapters/fs.js";

describe("FsAdapter", () => {
  test("adapter id is 'fs'", () => {
    const a = new FsAdapter();
    expect(a.id).toBe("fs");
  });

  test("fingerprint is deterministic", async () => {
    const a = new FsAdapter();
    const req: FsRequest = {
      op: "write",
      path: "/etc/hosts",
      data: "127.0.0.1 localhost\n",
    };
    const fp1 = await a.fingerprint(req);
    const fp2 = await a.fingerprint(req);
    expect(fp1).toHaveLength(8);
    expect(fp1).toBe(fp2);
  });

  test("fingerprint discriminates op", async () => {
    const a = new FsAdapter();
    const fpW = await a.fingerprint({ op: "write", path: "/x" });
    const fpR = await a.fingerprint({ op: "remove", path: "/x" });
    expect(fpW).not.toBe(fpR);
  });

  test("fingerprint discriminates path", async () => {
    const a = new FsAdapter();
    const fpA = await a.fingerprint({ op: "write", path: "/a", data: "x" });
    const fpB = await a.fingerprint({ op: "write", path: "/b", data: "x" });
    expect(fpA).not.toBe(fpB);
  });

  test("fingerprint discriminates data", async () => {
    const a = new FsAdapter();
    const fpA = await a.fingerprint({ op: "write", path: "/x", data: "foo" });
    const fpB = await a.fingerprint({ op: "write", path: "/x", data: "bar" });
    expect(fpA).not.toBe(fpB);
  });

  test("fingerprint discriminates mode", async () => {
    const a = new FsAdapter();
    const fpA = await a.fingerprint({
      op: "write",
      path: "/x",
      data: "y",
      mode: 0o644,
    });
    const fpB = await a.fingerprint({
      op: "write",
      path: "/x",
      data: "y",
      mode: 0o755,
    });
    expect(fpA).not.toBe(fpB);
  });

  test("fingerprint omits mode when undefined", async () => {
    const a = new FsAdapter();
    const bare = await a.fingerprint({ op: "write", path: "/x", data: "y" });
    const withUndef = await a.fingerprint({
      op: "write",
      path: "/x",
      data: "y",
      mode: undefined,
    });
    expect(bare).toBe(withUndef);
  });

  test("mode: 0 produces different fingerprint than mode undefined", async () => {
    // Matches Go's *uint32 semantics: pointer-to-zero vs nil.
    const a = new FsAdapter();
    const undef = await a.fingerprint({ op: "write", path: "/x", data: "y" });
    const zero = await a.fingerprint({
      op: "write",
      path: "/x",
      data: "y",
      mode: 0,
    });
    expect(undef).not.toBe(zero);
  });

  test("flags omitted when zero (matches Go's omitempty)", async () => {
    const a = new FsAdapter();
    const bare = await a.fingerprint({ op: "write", path: "/x", data: "y" });
    const withZeroFlags = await a.fingerprint({
      op: "write",
      path: "/x",
      data: "y",
      flags: 0,
    });
    expect(bare).toBe(withZeroFlags);
    const withFlags = await a.fingerprint({
      op: "write",
      path: "/x",
      data: "y",
      flags: 1,
    });
    expect(withFlags).not.toBe(bare);
  });

  test("recursive omitted when false", async () => {
    const a = new FsAdapter();
    const bare = await a.fingerprint({ op: "remove", path: "/x" });
    const withFalse = await a.fingerprint({
      op: "remove",
      path: "/x",
      recursive: false,
    });
    const withTrue = await a.fingerprint({
      op: "remove",
      path: "/x",
      recursive: true,
    });
    expect(bare).toBe(withFalse);
    expect(bare).not.toBe(withTrue);
  });

  test("dest omitted when empty", async () => {
    const a = new FsAdapter();
    const bare = await a.fingerprint({ op: "rename", path: "/a" });
    const withEmpty = await a.fingerprint({
      op: "rename",
      path: "/a",
      dest: "",
    });
    const withDest = await a.fingerprint({
      op: "rename",
      path: "/a",
      dest: "/b",
    });
    expect(bare).toBe(withEmpty);
    expect(bare).not.toBe(withDest);
  });

  test("data omitted when empty string", async () => {
    const a = new FsAdapter();
    const bare = await a.fingerprint({ op: "write", path: "/x" });
    const empty = await a.fingerprint({ op: "write", path: "/x", data: "" });
    expect(bare).toBe(empty);
  });

  // CRITICAL: conformance fixture — locks the cross-runtime contract.
  test("conformance: fs-write fixture fingerprint equals 667a7680", async () => {
    const a = new FsAdapter();
    const fp = await a.fingerprint({
      op: "write",
      path: "$TMP/greeting.txt",
      data: "hello, world\n",
      mode: 420,
    });
    expect(fp).toBe("667a7680");
  });

  test("serialize/deserialize round-trip", () => {
    const a = new FsAdapter();
    const req: FsRequest = {
      op: "write",
      path: "/etc/hosts",
      data: "127.0.0.1 localhost\n",
      mode: 0o644,
    };
    const ser = a.serializeReq(req);
    const got = a.deserializeReq(ser);
    expect(got).toEqual(req);
  });

  test("response serialize/deserialize round-trip", () => {
    const a = new FsAdapter();
    const resp = { duration_ms: 42, bytes_written: 1024 };
    const ser = a.serializeResp(resp);
    const got = a.deserializeResp(ser);
    expect(got).toEqual(resp);
  });

  test("base64 payload round-trips byte-exact through YAML", async () => {
    // Spec "Data Field Encoding": binary callers base64-encode before the
    // adapter sees `data`; the adapter and the YAML layer treat the string
    // as opaque; the caller decodes on the way back.
    const a = new FsAdapter();
    const raw = Buffer.from([0x00, 0xff, 0xc3, 0x28, 0x80, 0x01, 0x02, 0x03]);
    const encoded = raw.toString("base64");
    const req: FsRequest = { op: "write", path: "/bin/x", data: encoded };

    const text = yaml.dump(a.serializeReq(req));
    expect(text).toContain(encoded);
    expect(text).not.toContain("!!binary");

    const got = a.deserializeReq(yaml.load(text));
    expect(got.data).toBe(encoded);
    expect(Buffer.from(got.data ?? "", "base64")).toEqual(raw);
    // Opaque to the fingerprint too: text or base64, only the bytes matter.
    expect(await a.fingerprint(got)).toBe(await a.fingerprint(req));
  });
});

describe("FsAdapter path normalizer", () => {
  const tmpRoot = "/var/folders/abc/T/run-123";
  const stripTmp: PathNormalizer = (p) => p.replace(tmpRoot, "$TMP");

  test("default normalizer is identity", () => {
    const a = new FsAdapter();
    expect(a.normalize("/var/folders/abc/T/run-123/x")).toBe(
      "/var/folders/abc/T/run-123/x"
    );
  });

  test("normalize() rewrites paths and is idempotent for prefix rules", () => {
    const a = new FsAdapter().withNormalizer(stripTmp);
    expect(a.normalize(`${tmpRoot}/x`)).toBe("$TMP/x");
    expect(a.normalize("$TMP/x")).toBe("$TMP/x");
  });

  test("normalizer is applied to path before fingerprinting", async () => {
    const plain = new FsAdapter();
    const norm = new FsAdapter().withNormalizer(stripTmp);
    const raw: FsRequest = {
      op: "write",
      path: `${tmpRoot}/config.yaml`,
      data: "k: v",
    };
    const pre: FsRequest = { op: "write", path: "$TMP/config.yaml", data: "k: v" };

    const fpRawNorm = await norm.fingerprint(raw);
    const fpPreNorm = await norm.fingerprint(pre);
    expect(fpRawNorm).toBe(fpPreNorm);

    const fpRawPlain = await plain.fingerprint(raw);
    expect(fpRawPlain).not.toBe(fpRawNorm);
  });

  test("two tmp roots collapse to one fingerprint when the root is stripped", async () => {
    const rootA = "/var/folders/aa/T/run-1";
    const rootB = "/private/var/folders/bb/T/run-2";
    const a = new FsAdapter().withNormalizer((p) => p.replace(rootA, "$TMP"));
    const b = new FsAdapter().withNormalizer((p) => p.replace(rootB, "$TMP"));
    const fpA = await a.fingerprint({ op: "write", path: `${rootA}/out.txt`, data: "x" });
    const fpB = await b.fingerprint({ op: "write", path: `${rootB}/out.txt`, data: "x" });
    expect(fpA).toBe(fpB);
  });

  test.each<FsRequest["op"]>(["rename", "symlink", "hardlink"])(
    "normalizer is applied to dest for %s",
    async (op) => {
      const plain = new FsAdapter();
      const norm = new FsAdapter().withNormalizer(stripTmp);
      const raw: FsRequest = { op, path: `${tmpRoot}/a`, dest: `${tmpRoot}/b` };
      const pre: FsRequest = { op, path: "$TMP/a", dest: "$TMP/b" };
      expect(await norm.fingerprint(raw)).toBe(await norm.fingerprint(pre));
      expect(await plain.fingerprint(raw)).not.toBe(await norm.fingerprint(raw));
    }
  );

  test("empty path and dest short-circuit without invoking the normalizer", async () => {
    let calls = 0;
    const a = new FsAdapter().withNormalizer(() => {
      calls++;
      return "NEVER";
    });
    expect(a.normalize("")).toBe("");
    await a.fingerprint({ op: "chmod", path: "", mode: 0o644 });
    const ser = a.serializeReq({ op: "rename", path: "", dest: "" }) as FsRequest;
    expect(ser.path).toBe("");
    expect(ser.dest).toBe("");
    expect(calls).toBe(0);
  });

  test("dest is gated on the normalized value, not the raw one", async () => {
    // spec: `dest` participates only when non-empty AFTER normalization.
    const drop: PathNormalizer = (p) => (p === "/x/drop" ? "" : p);
    const a = new FsAdapter().withNormalizer(drop);
    const noDest = await a.fingerprint({ op: "rename", path: "/a" });
    const dropped = await a.fingerprint({ op: "rename", path: "/a", dest: "/x/drop" });
    const kept = await a.fingerprint({ op: "rename", path: "/a", dest: "/x/keep" });
    expect(dropped).toBe(noDest);
    expect(kept).not.toBe(noDest);
  });

  test("empty dest stays omitted regardless of the normalizer", async () => {
    const a = new FsAdapter().withNormalizer((p) => (p === "" ? "/ghost" : p));
    const noDest = await a.fingerprint({ op: "rename", path: "/a" });
    const empty = await a.fingerprint({ op: "rename", path: "/a", dest: "" });
    expect(empty).toBe(noDest);
  });

  test("chainNormalizers composes left to right", () => {
    const tmpToPlaceholder: PathNormalizer = (p) => p.replace("/tmp", "$TMP");
    const placeholderToShort: PathNormalizer = (p) => p.replace("$TMP", "$T");
    expect(chainNormalizers(tmpToPlaceholder, placeholderToShort)("/tmp/x")).toBe("$T/x");
    expect(chainNormalizers(placeholderToShort, tmpToPlaceholder)("/tmp/x")).toBe("$TMP/x");
    expect(chainNormalizers()("/tmp/x")).toBe("/tmp/x");
  });

  test("chained normalizer drives the fingerprint", async () => {
    const tmpNorm: PathNormalizer = (p) => p.replace("/tmp", "$TMP");
    const homeNorm: PathNormalizer = (p) => p.replace("/home/u", "$HOME");
    const a = new FsAdapter().withNormalizer(chainNormalizers(tmpNorm, homeNorm));
    const fp1 = await a.fingerprint({ op: "rename", path: "/tmp/foo", dest: "/home/u/bar" });
    const fp2 = await a.fingerprint({ op: "rename", path: "$TMP/foo", dest: "$HOME/bar" });
    expect(fp1).toBe(fp2);
  });

  test("serializeReq stores post-normalizer path and dest", () => {
    const a = new FsAdapter().withNormalizer(stripTmp);
    const req: FsRequest = {
      op: "rename",
      path: `${tmpRoot}/old`,
      dest: `${tmpRoot}/new`,
      mode: 0o644,
    };
    const out = a.serializeReq(req) as FsRequest;
    expect(out).toEqual({
      op: "rename",
      path: "$TMP/old",
      dest: "$TMP/new",
      mode: 0o644,
    });
    expect(JSON.stringify(out)).not.toContain(tmpRoot);
    // Caller's request object is left untouched.
    expect(req.path).toBe(`${tmpRoot}/old`);
    expect(req.dest).toBe(`${tmpRoot}/new`);
  });

  test("serializeReq does not introduce a dest key when absent", () => {
    const a = new FsAdapter().withNormalizer(stripTmp);
    const out = a.serializeReq({ op: "write", path: `${tmpRoot}/f`, data: "x" });
    expect(Object.keys(out as object)).toEqual(["op", "path", "data"]);
  });

  test("replay-side plain adapter re-derives the recorded fingerprint from the stored request", async () => {
    // Cross-runtime contract: the cassette carries the post-normalizer
    // path, so a replay side with NO normalizer (any other port) must
    // hash the stored request to the same fingerprint the recorder used.
    const recorder = new FsAdapter().withNormalizer(stripTmp);
    const replayer = new FsAdapter();
    const raw: FsRequest = { op: "rename", path: `${tmpRoot}/a`, dest: `${tmpRoot}/b` };
    const stored = replayer.deserializeReq(recorder.serializeReq(raw));
    expect(await replayer.fingerprint(stored)).toBe(await recorder.fingerprint(raw));
  });

  test("withNormalizer returns a new adapter and leaves the original untouched", async () => {
    const base = new FsAdapter();
    const norm = base.withNormalizer(stripTmp);
    expect(norm).not.toBe(base);
    expect(norm.id).toBe("fs");
    const raw: FsRequest = { op: "write", path: `${tmpRoot}/f`, data: "x" };
    expect(await base.fingerprint(raw)).not.toBe(await norm.fingerprint(raw));
    expect(base.normalize(`${tmpRoot}/f`)).toBe(`${tmpRoot}/f`);
  });

  test("constructor option is equivalent to withNormalizer", async () => {
    const viaCtor = new FsAdapter({ normalizer: stripTmp });
    const viaWith = new FsAdapter().withNormalizer(stripTmp);
    const raw: FsRequest = { op: "write", path: `${tmpRoot}/f`, data: "x" };
    expect(await viaCtor.fingerprint(raw)).toBe(await viaWith.fingerprint(raw));
  });

  test("conformance: normalized tmp path reproduces the fs-write pin 667a7680", async () => {
    const a = new FsAdapter().withNormalizer(stripTmp);
    const fp = await a.fingerprint({
      op: "write",
      path: `${tmpRoot}/greeting.txt`,
      data: "hello, world\n",
      mode: 420,
    });
    expect(fp).toBe("667a7680");
  });
});

File: go/adapters/fs/fs_test.go
package fs_test

import (
	"encoding/base64"
	"path/filepath"
	"strings"
	"testing"

	xrr "hop.top/xrr"
	"hop.top/xrr/adapters/fs"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// Task 1: identity tests.

func TestAdapterID(t *testing.T) {
	a := fs.NewAdapter()
	assert.Equal(t, "fs", a.ID())
}

func TestRequestAdapterID(t *testing.T) {
	r := &fs.Request{Op: fs.OpWrite, Path: "/x"}
	assert.Equal(t, "fs", r.AdapterID())
}

func TestResponseAdapterID(t *testing.T) {
	r := &fs.Response{}
	assert.Equal(t, "fs", r.AdapterID())
}

// Task 2: fingerprint tests.

func TestFingerprintDeterministic(t *testing.T) {
	a := fs.NewAdapter()
	req := &fs.Request{Op: fs.OpWrite, Path: "/etc/hosts", Data: "127.0.0.1 localhost\n"}
	fp1, err := a.Fingerprint(req)
	require.NoError(t, err)
	assert.Len(t, fp1, 8, "fingerprint must be 8 hex chars")
	fp2, err := a.Fingerprint(req)
	require.NoError(t, err)
	assert.Equal(t, fp1, fp2, "same request must hash identically")
}

func TestFingerprintDiscriminatesPath(t *testing.T) {
	a := fs.NewAdapter()
	fpA, _ := a.Fingerprint(&fs.Request{Op: fs.OpWrite, Path: "/a", Data: "x"})
	fpB, _ := a.Fingerprint(&fs.Request{Op: fs.OpWrite, Path: "/b", Data: "x"})
	assert.NotEqual(t, fpA, fpB)
}

func TestFingerprintDiscriminatesData(t *testing.T) {
	a := fs.NewAdapter()
	fpA, _ := a.Fingerprint(&fs.Request{Op: fs.OpWrite, Path: "/x", Data: "foo"})
	fpB, _ := a.Fingerprint(&fs.Request{Op: fs.OpWrite, Path: "/x", Data: "bar"})
	assert.NotEqual(t, fpA, fpB)
}

func TestFingerprintDiscriminatesOp(t *testing.T) {
	a := fs.NewAdapter()
	fpW, _ := a.Fingerprint(&fs.Request{Op: fs.OpWrite, Path: "/x"})
	fpR, _ := a.Fingerprint(&fs.Request{Op: fs.OpRemove, Path: "/x"})
	assert.NotEqual(t, fpW, fpR)
}

func TestFingerprintOmitsZeroFields(t *testing.T) {
	a := fs.NewAdapter()
	// Request with Mode unset must hash identically to a "minimal" write.
	bare := &fs.Request{Op: fs.OpWrite, Path: "/x", Data: "y"}
	withNilMode := &fs.Request{Op: fs.OpWrite, Path: "/x", Data: "y", Mode: nil}
	fpA, _ := a.Fingerprint(bare)
	fpB, _ := a.Fingerprint(withNilMode)
	assert.Equal(t, fpA, fpB, "Mode: nil must omit `mode` from fingerprint")
}

func TestFingerprintPointerToZeroIncludesField(t *testing.T) {
	a := fs.NewAdapter()
	zero := uint32(0)
	bare := &fs.Request{Op: fs.OpWrite, Path: "/x", Data: "y"}
	withZeroMode := &fs.Request{Op: fs.OpWrite, Path: "/x", Data: "y", Mode: &zero}
	fpA, _ := a.Fingerprint(bare)
	fpB, _ := a.Fingerprint(withZeroMode)
	assert.NotEqual(t, fpA, fpB,
		"Mode: &0 must include `mode: 0` in fingerprint; differs from Mode: nil")
}

func TestFingerprintRejectsWrongType(t *testing.T) {
	a := fs.NewAdapter()
	_, err := a.Fingerprint(notFsRequest{})
	assert.Error(t, err)
}

type notFsRequest struct{}

func (notFsRequest) AdapterID() string { return "not-fs" }

// TestFingerprintConformanceFsWrite — cross-runtime conformance: the
// spec/fixtures/fs-write request MUST hash to 667a7680, the value the
// TypeScript, Python, Rust, and PHP ports each pin. The request is loaded
// through the cassette rather than hand-built so the fixture bytes are
// pinned too: a drift in either the fixture or the fingerprint fails here.
func TestFingerprintConformanceFsWrite(t *testing.T) {
	dir := filepath.Join("..", "..", "..", "spec", "fixtures", "fs-write")
	var req fs.Request
	var resp fs.Response
	_, err := xrr.NewFileCassette(dir).Load("fs", "667a7680", &req, &resp)
	require.NoError(t, err)

	assert.Equal(t, fs.OpWrite, req.Op)
	assert.Equal(t, "$TMP/greeting.txt", req.Path)
	assert.Equal(t, "hello, world\n", req.Data)
	require.NotNil(t, req.Mode)
	assert.Equal(t, uint32(420), *req.Mode) // 0o644

	fp, err := fs.NewAdapter().Fingerprint(&req)
	require.NoError(t, err)
	assert.Equal(t, "667a7680", fp, "spec conformance fingerprint mismatch")
}

// Task 3: PathNormalizer tests.

func TestNormalizerAppliedToFingerprint(t *testing.T) {
	plain := fs.NewAdapter()
	norm := fs.NewAdapter().WithNormalizer(func(p string) string {
		return strings.Replace(p, "/var/folders/abc/T/Test123", "$TMP", 1)
	})

	rawReq := &fs.Request{Op: fs.OpWrite, Path: "/var/folders/abc/T/Test123/config.yaml", Data: "k: v"}
	normReq := &fs.Request{Op: fs.OpWrite, Path: "$TMP/config.yaml", Data: "k: v"}

	// Plain adapter sees raw path; normalized adapter sees rewritten path.
	// The two fingerprints from the NORMALIZED adapter must match —
	// regardless of which raw path was fed in — because both reduce
	// to the same normalized form.
	fpRawNorm, _ := norm.Fingerprint(rawReq)
	fpNormNorm, _ := norm.Fingerprint(normReq)
	assert.Equal(t, fpRawNorm, fpNormNorm,
		"normalizer must map both raw and pre-normalized paths to same fp")

	// And the plain adapter MUST disagree with the normalizing adapter
	// on the raw request, proving the normalizer actually ran.
	fpRawPlain, _ := plain.Fingerprint(rawReq)
	assert.NotEqual(t, fpRawPlain, fpRawNorm,
		"plain adapter and normalizing adapter must differ on raw path input")
}

func TestNormalizerAppliedToDest(t *testing.T) {
	norm := fs.NewAdapter().WithNormalizer(func(p string) string {
		return strings.Replace(p, "/tmp", "$TMP", 1)
	})
	req := &fs.Request{Op: fs.OpRename, Path: "/tmp/a", Dest: "/tmp/b"}
	fp1, _ := norm.Fingerprint(req)
	// Same request with already-normalized paths must produce same fp.
	req2 := &fs.Request{Op: fs.OpRename, Path: "$TMP/a", Dest: "$TMP/b"}
	fp2, _ := norm.Fingerprint(req2)
	assert.Equal(t, fp1, fp2)
}

func TestChainNormalizer(t *testing.T) {
	tmpNorm := func(p string) string { return strings.Replace(p, "/tmp", "$TMP", 1) }
	homeNorm := func(p string) string { return strings.Replace(p, "/home/u", "$HOME", 1) }
	a := fs.NewAdapter().WithNormalizer(fs.Chain(tmpNorm, homeNorm))

	req := &fs.Request{Op: fs.OpWrite, Path: "/tmp/foo", Data: "x"}
	req2 := &fs.Request{Op: fs.OpWrite, Path: "$TMP/foo", Data: "x"}
	fp1, _ := a.Fingerprint(req)
	fp2, _ := a.Fingerprint(req2)
	assert.Equal(t, fp1, fp2, "chained normalizer must compose left to right")
}

func TestNormalizerEmptyPathPassesThrough(t *testing.T) {
	// An empty Path is invalid per the wrapper, but the adapter
	// itself should not crash. The normalize() helper must short-
	// circuit on "" without invoking the user's function.
	calls := 0
	a := fs.NewAdapter().WithNormalizer(func(p string) string {
		calls++
		return "NEVER"
	})
	// Use chmod with no path field set (test of robustness only).
	mode := uint32(0o644)
	_, err := a.Fingerprint(&fs.Request{Op: fs.OpChmod, Path: "", Mode: &mode})
	require.NoError(t, err)
	assert.Equal(t, 0, calls, "empty path must not invoke normalizer")
}

// TestDestGatedOnNormalizedValue — spec: `dest` participates only
// when non-empty AFTER path normalization. A normalizer that maps a
// non-empty dest to "" drops it from the fingerprint, so the request
// hashes identically to one with no dest at all. The no-dest hash is
// the cross-port vector sha256(`{"op":"rename","path":"/a"}`)[:8].
func TestDestGatedOnNormalizedValue(t *testing.T) {
	a := fs.NewAdapter().WithNormalizer(func(p string) string {
		if p == "/x/drop" {
			return ""
		}
		return p
	})
	noDest := &fs.Request{Op: fs.OpRename, Path: "/a"}
	dropped := &fs.Request{Op: fs.OpRename, Path: "/a", Dest: "/x/drop"}
	kept := &fs.Request{Op: fs.OpRename, Path: "/a", Dest: "/x/keep"}

	fpNoDest, err := a.Fingerprint(noDest)
	require.NoError(t, err)
	fpDropped, err := a.Fingerprint(dropped)
	require.NoError(t, err)
	fpKept, err := a.Fingerprint(kept)
	require.NoError(t, err)

	assert.Equal(t, "86f75341", fpNoDest, "cross-port no-dest vector")
	assert.Equal(t, fpNoDest, fpDropped,
		"dest normalized to \"\" must drop out of the fingerprint")
	assert.NotEqual(t, fpNoDest, fpKept,
		"dest that survives normalization must still discriminate")

	// The persisted copy agrees: a wrapper that pre-normalizes Dest
	// (per the PathNormalizer contract) stores no `dest:` at all.
	stored := &fs.Request{Op: fs.OpRename, Path: "/a", Dest: a.Normalize("/x/drop")}
	data, err := a.Serialize(stored)
	require.NoError(t, err)
	assert.NotContains(t, string(data), "dest:")
}

// TestEmptyDestStaysOmittedRegardlessOfNormalizer — an unset Dest
// never reaches the normalizer (Normalize short-circuits on ""), so a
// normalizer that would rewrite "" cannot conjure a dest into the
// fingerprint.
func TestEmptyDestStaysOmittedRegardlessOfNormalizer(t *testing.T) {
	a := fs.NewAdapter().WithNormalizer(func(p string) string {
		if p == "" {
			return "/ghost"
		}
		return p
	})
	fpNoDest, err := a.Fingerprint(&fs.Request{Op: fs.OpRename, Path: "/a"})
	require.NoError(t, err)
	fpEmpty, err := a.Fingerprint(&fs.Request{Op: fs.OpRename, Path: "/a", Dest: ""})
	require.NoError(t, err)
	assert.Equal(t, fpNoDest, fpEmpty)
}

// Task 4: Serialize/Deserialize tests.

func TestSerializeRoundtrip(t *testing.T) {
	a := fs.NewAdapter()
	mode := uint32(0o644)
	req := &fs.Request{
		Op:    fs.OpWrite,
		Path:  "/etc/hosts",
		Data:  "127.0.0.1 localhost\n",
		Mode:  &mode,
		Flags: 0,
	}
	data, err := a.Serialize(req)
	require.NoError(t, err)
	var got fs.Request
	require.NoError(t, a.Deserialize(data, &got))
	assert.Equal(t, req.Op, got.Op)
	assert.Equal(t, req.Path, got.Path)
	assert.Equal(t, req.Data, got.Data)
	require.NotNil(t, got.Mode)
	assert.Equal(t, *req.Mode, *got.Mode)
}

func TestSerializeBase64Payload(t *testing.T) {
	// Caller base64-encodes binary data before passing to the adapter.
	// The adapter serializes it as an opaque UTF-8 string. Round-trip
	// recovers the exact string; the caller is responsible for
	// base64-decoding back to bytes.
	a := fs.NewAdapter()
	rawBinary := []byte{0x00, 0xff, 0xc3, 0x28, 0x80, 0x01, 0x02, 0x03}
	encoded := base64.StdEncoding.EncodeToString(rawBinary)
	req := &fs.Request{Op: fs.OpWrite, Path: "/bin/x", Data: encoded}
	data, err := a.Serialize(req)
	require.NoError(t, err)
	t.Logf("YAML output:\n%s", string(data))
	var got fs.Request
	require.NoError(t, a.Deserialize(data, &got))
	assert.Equal(t, encoded, got.Data, "base64 string must round-trip exactly")
	// And caller can recover the original bytes.
	decoded, err := base64.StdEncoding.DecodeString(got.Data)
	require.NoError(t, err)
	assert.Equal(t, rawBinary, decoded)
}

func TestSerializeOmitsZeroOptionals(t *testing.T) {
	// A bare write request must not emit `dest:`, `mode:`, `uid:`,
	// `gid:`, `size:`, `flags:`, or `recursive:` in YAML.
	a := fs.NewAdapter()
	req := &fs.Request{Op: fs.OpWrite, Path: "/x", Data: "y"}
	data, _ := a.Serialize(req)
	out := string(data)
	for _, forbidden := range []string{"dest:", "mode:", "uid:", "gid:", "size:", "flags:", "recursive:"} {
		assert.NotContains(t, out, forbidden,
			"bare write request must omit %q from YAML", forbidden)
	}
}

func TestSerializeResponseRoundtrip(t *testing.T) {
	a := fs.NewAdapter()
	resp := &fs.Response{DurationMs: 42, BytesWritten: 1024}
	data, err := a.Serialize(resp)
	require.NoError(t, err)
	var got fs.Response
	require.NoError(t, a.Deserialize(data, &got))
	assert.Equal(t, resp.DurationMs, got.DurationMs)
	assert.Equal(t, resp.BytesWritten, got.BytesWritten)
}

// TestNormalizeIsPublicAndIdempotent — wrappers call Adapter.Normalize
// when building Request.Path / Request.Dest so the persisted
// cassette payload agrees with what Fingerprint hashes. Regression
// test for the spec contract "cassettes store post-normalizer paths"
// (without this, Fingerprint normalizes but Serialize persists raw).
func TestNormalizeIsPublicAndIdempotent(t *testing.T) {
	a := fs.NewAdapter().WithNormalizer(func(p string) string {
		return strings.Replace(p, "/tmp/run-123", "$TMP", 1)
	})

	// Public API: wrappers can call Normalize at request construction.
	assert.Equal(t, "$TMP/x", a.Normalize("/tmp/run-123/x"))
	assert.Equal(t, "$TMP/x", a.Normalize("$TMP/x"),
		"normalizing an already-normalized path is a no-op for this rule")
	assert.Equal(t, "", a.Normalize(""),
		"empty path short-circuits without invoking the normalizer")
}

// TestSerializeStoresPostNormalizerPath — when a wrapper has
// pre-normalized Path/Dest before constructing the Request, the
// YAML payload written to disk MUST contain the normalized form.
// This is the contract cross-runtime replay relies on.
func TestSerializeStoresPostNormalizerPath(t *testing.T) {
	a := fs.NewAdapter().WithNormalizer(func(p string) string {
		return strings.Replace(p, "/tmp/run-123", "$TMP", 1)
	})

	// Wrappers MUST do this normalization at construction time.
	req := &fs.Request{
		Op:   fs.OpRename,
		Path: a.Normalize("/tmp/run-123/old"),
		Dest: a.Normalize("/tmp/run-123/new"),
	}

	data, err := a.Serialize(req)
	require.NoError(t, err)
	out := string(data)
	assert.Contains(t, out, "path: $TMP/old")
	assert.Contains(t, out, "dest: $TMP/new")
	assert.NotContains(t, out, "/tmp/run-123",
		"serialized payload must not leak the raw tmpdir path")
}

// Cross-port hazard vector (spec: Fingerprint Algorithm). The same path
// pins the same fingerprint in every port.
func TestFingerprintHazardVector(t *testing.T) {
	hazard := "a&b<c>/é" + string(rune(0x2028)) + string(rune(0x2029)) + "\b\f\x1f\x7f"
	fp, err := fs.NewAdapter().Fingerprint(&fs.Request{Op: "write", Path: hazard})
	require.NoError(t, err)
	assert.Equal(t, "6f2fb087", fp)
}
